stdlib: an indirection for a memory-safe buffer

Octachron opened the pull request:
In OCaml 4, the buffer module was using internally unsafe bytes accesses. Those accesses relied on invariants directly dependent on the global runtime lock to ensure safety even in case of concurrent uses of the same buffer.

Those invariants no longer holds in OCaml 5. As a consequence, it is no longer possible to use unsafe accesses as liberally as in OCaml 4. A straightforward way to restore memory safety is to give up on all unsafe operations as proposed in #11585 .

To reach a compromise between performance safety,  @lthls  proposed in #11279 proposed to add a new level of indirection to the buffer data type in order to keep a fast path whenever the buffer does not need resizing and fall back to safe `Bytes` function otherwise.

More precisely, this PR changes the memory representation of buffer by replacing the two mutable [length] and [buffer] fields by a single mutable [inner] field which holds an immutable inner buffer.

With this new indirection,  we can enforce that for a inner buffer `ib`, `ib.length` = `Bytes.length ib.buffer`.
Performance sensitive functions uses unsafe operation of this immutable buffer whenever they don't need to extend the inner buffer themselves. Those "performance-sensitive" are currently:

- add_string
- add_char
- add_substring
- add_int8*
- add_int16*
- add_int32*
- add_int64*

In term of performance, I have collected a rough benchmark at https://github.com/Octachron/ocaml_buffer_benchmarks (see also https://github.com/dra27/ocaml_buffer_benchmarks for Windows data).

This benchmark tests buffer implementation on 5 scenarios: 
- a hot loop adding characters
-  a hot loop adding strings
-  the fasta3 test from the sandmark benchmarks modified to use only buffer rather than `Format`
-  a recursive tree printing code
-  a n-gram Markov chain text generator.

The two first scenarios are parameterized by the number of addition before a buffer reset and the number of resets.
The last three scenarios are parameterized by the size of the output.

This benchmark compares 4 different safe implementation:

- the one proposed in this PR (`indirection`)
-  a version of buffer without the `length` field optimisation (`simplified`) 
- the current buffer implementation `std`.
-  the `safe` implementation of  #11585 
All those implementations except for `std` are memory-sage.

To make the comparison between different implementations easier to read, I have collected the number time each of those implementations have been ranked first, second or third when comparing the median run time for sets of 100 or 1000 runs of the five scenario described with varying parameters.

|                     | 1st rank | 2nd rank | 3rd rank | other |
|-----------------|----------|----------|----------|-------|
| indirection| 68.7%    | 19.4%    | 11.2%    | 0.7% |
| simplified  | 13.3%    | 45.7%    | 32.0%    | 9.0% |
| std              | 17.3%    | 30.2%    | 44.2%    | 8.3% |
| safe            |  0.7%     | 4.7%    | 12.6%    | 82.0% |

(on a total of 278 tests).

Looking at other quantiles doesn't make much different to those tables.
As shown in this table, the `indirection` implementation is most of the time the fastest implementation and is very rarely the slowest implementation. 

Restricting the comparison between the `std` and `indirection` implementations  indicates that the median run time of this PR `indirection` implementation is shorter than the one of the `std` implementation in nearly 80% of cases:

|             | 1st rank | 2nd rank |
|-------------|----------|----------|
| indirection        | 79.5%    | 20.5%    | 
| std         | 20.5%    | 79.5%    | 


Moreover, the case where the `indirection` test is the slowest often corresponds to the short runtime parameters which are quite noisy.

Overall, I think that the implementation proposed in this PR represents a decent compromise in term of performance and complexity of the code for a memory-safe implementation of buffer.

Octachron committed:
An indirection for a memory-safe buffer implementation

In OCaml 4, the buffer module was using internally unsafe bytes accesses.
Those accesses relied on invariants directly dependent on the global runtime
lock to ensure safety even in case of concurrent uses of the same buffer.

Those invariants no longer holds in OCaml 5. As a consequence, it is no longer
possible to use unsafe accesses as liberally as in OCaml 4.

To reach a compromise between performance safety, this commit changes
the buffer implementation to fall back on safe functions in most cases,
but adds a fast path to performance sensitive functions whenever it is possible
to guarantee memory safety even in presence of data races (while offfering no
correctness guarantees).

More precisely, this commit changes the memory representation of buffer
by replacing the two mutable [length] and [buffer] fields by a single
mutable [inner] field which holds an immutable inner buffer.

This change makes it possible to enforece that for a inner buffer
[ib.length = Bytes.length ib.buffer].
Performance sensitive functions uses unsafe operation of this immutable
buffer whenever they don't need to mutate the inner buffer themselves.
Those "performance-sensitive" are currently:

- add_string
- add_char
- add_substring
- add_int8*
- add_int16*
- add_int32*
- add_int64*

gasche commented on stdlib/buffer.ml:
```diff
@@ -87,102 +96,91 @@ let resize b more =
+    Bytes.set b.inner.buffer pos c
+  ) else
+    Bytes.unsafe_set buffer pos c;
   b.position <- pos + 1
```
I would write a simpler/shorter version:

```ocaml
let rec add_char b c =
  let pos = b.position in
  let {buffer;length} = b.inner in
  if pos >= length then (
    resize b 1;
    add_char b c
  ) else (
    b.position <- pos + 1;
    Bytes.unsafe_set buffer pos c;
  )
```

lthls replied:
My own take on this (and the other functions) would be to make the basic `resize` function return the inner buffer (with its length) instead of setting it. This way you can still use unsafe accesses, and you don't have to worry about add_char looping infinitely if run in parallel with `reset` operations. You could even share the code for updating the buffer and the position, although that will likely cause a small performance hit because of the introduction of spill/reload instructions from the cold path into the hot path.

Octachron replied:
After a quick glance, the benefits of both the recursive and the `resize-once` variants are not obvious in term of performance:

|                               | 1st rank | 2nd rank | 3rd rank | other |
|-------------------------------|----------|----------|----------|-------|
| indirection                   | 20.2%    | 46.4%    | 19.0%    | 14.3% |
| indirection+rec               | 32.1%    | 19.0%    | 36.9%    | 11.9% |
| indirection+resize-once       | 44.0%    | 16.7%    | 35.7%    | 3.6% |
| indirection+resize-once+spill | 1.2%    | 11.9%    | 4.8%    | 82.1% |
| std                           | 2.4%    | 6.0%    | 3.6%    | 88.1% |

- `indirection`: this PR
- `indirection+rec`: @gasche proposition
- `indirection+resize-once`: @lthls proposition
- `indirection+resize-once+spill` : @lthls proposition with maximal code sharing between branches
- `std` : current stdlib implementation

(This is on my linux machine that was favorable to the `indirection` test in the earlier results)

Restraining the comparison to the fast variants of the current implementation gives me the impression that there is no variant really faster than the other:

|                               | 1st rank | 2nd rank | 3rd rank |
|-------------------------------|----------|----------|----------|
| indirection+resize-once       | 44.0%    | 16.7%    | 39.3%    | 
| indirection+rec               | 31.0%    | 29.8%    | 39.3%    | 
| indirection                   | 25.0%    | 53.6%    | 21.4%    |

Note that this is on the `medium` variant of the benchmark which might be sensitive to the noise. I will share the result of the larger variant once the run is done later this evening.

gasche replied:
My recursive add_char is not supposed to be faster (it should have basically the same performance), but to be simpler (to the human reader; not having a safe and unsafe instructions close to each other) and shorter (in term of generated code size, I expect a sizeable reduction). Code-size effects tend to come back to bite us from time to time (weird code-cache questions in real-world large programs, but also js_of_ocaml etc.), so it's something to keep in mind.

@lthls's suggestion could also be nice as long as it is also simpler and shorter. (The easiest/nicest way to write it is with a `let buffer = ensure_room b.inner pos 1 in` function, but I would not expect this to be performance-preserving because we either pay a function call in the hot path or inline a lot more code. The current code structure, and even more the recursive version, are good at hiding the cold path away.)

lthls replied:
I've thought of nice ways to make the code both short and fast, and my favourite solution looks like that:
```ocaml
let add_char b c =
  let pos = b.position in
  let[@local_or_inline] update inner =
    Bytes.unsafe_set inner.buffer pos c;
    b.position <- pos + 1
  in
  let inner = b.inner in
  if pos >= inner.length then (
    let new_inner = resize b 1 in
    update new_inner
  ) else
    update inner
``` 
However I'd need to have a new `[@local_or_inline]` annotation that means `[@local]` in bytecode, to avoid extra function calls, but `[@inline]` in native mode, to keep separate hot and cold paths, so I can't implement it currently.

Note that the use of recursive functions is going to change the inlining heuristics significantly. The inlining heuristics for recursive functions are much more conservative than for non-recursive ones (in flambda at least), so I assume that this will impact performance negatively. Adding an `[@inline]` annotation on the functions that we want inlined should solve this, but it's not very satisfying.

Octachron replied:
@gasche , the size of the object files don't seem to match your expectation?

- indirection.o:  35912
- indirection_rec.o: 36376 
- indirection_resize_once.o: 37768
- indirection_resize_once_spill.o: 38400

I am also uncertain if the recursive variant is simpler. In particular, the fact that this variant is unbounded in time seem to add some complexity to this solution, compared to for instance

```ocaml

let add_char b c =
  let pos = b.position in
  let {buffer;length} = b.inner in
  if pos >= length then begin
    let buffer = extended_inner_buffer b 1 in
    Bytes.unsafe_set buffer pos c;
    b.position <- pos + 1
    end
 else begin
    Bytes.unsafe_set buffer pos c;
    b.position <- pos + 1
  end
```
which is visibly using the same unsafe function.

gasche replied:
I'm starting to regret this thread of conversation -- guilty. Why don't we just take @Octachron's code and run with it?

gasche replied:
@Octachron on files trimmed to contain just the type declarations, `resize` as a C external and `add_char`, on my machine I can observe a size reduction for the recursive version (.text is the code segment):

```
$ size --format=GNU add.o addrec.o
      text       data        bss      total filename
       199        320          0        519 add.o
       135        344          0        479 addrec.o
```

add.ml:

```ocaml
type inner_buffer = {
  buffer: bytes;
  length: int;
}

type t =
 {mutable inner : inner_buffer;
  mutable position : int;
  initial_buffer : bytes}

external resize : t -> int -> unit = "resize"

let add_char b c =
  let pos = b.position in
  let {buffer;length} = b.inner in
  if pos >= length then (
    resize b 1;
    Bytes.set b.inner.buffer pos c
  ) else
    Bytes.unsafe_set buffer pos c;
  b.position <- pos + 1
```

addrec.ml:

```ocaml
type inner_buffer = {
  buffer: bytes;
  length: int;
}
type t =
 {mutable inner : inner_buffer;
  mutable position : int;
  initial_buffer : bytes}

external resize : t -> int -> unit = "resize"

let rec add_char b c =
  let pos = b.position in
  let {buffer;length} = b.inner in
  if pos >= length then (
    resize b 1;
    add_char b c
  ) else (
    Bytes.unsafe_set buffer pos c;
    b.position <- pos + 1;
  )
```

Octachron replied:
You are right,  comparing the code segment shows that there is a small decrease in size for the rec version:

```
      text       data        bss      total filename
      8554       6648          0      15202 indirection.o
      8490       6744          0      15234 indirection_rec.o
      8670       6752          0      15422 indirection_resize_once.o
      8542       6840          0      15382 indirection_resize_once_spill.o
```

gasche commented on stdlib/buffer.ml:
```diff
@@ -87,102 +96,91 @@ let resize b more =
+    Bytes.blit_string s offset b.inner.buffer b.position len
+  ) else
+    Bytes.unsafe_blit_string s offset buffer position len;
   b.position <- new_position
```
same here, we could perform a recursive call on resize

gasche commented on stdlib/buffer.ml:
```diff
@@ -87,102 +96,91 @@ let resize b more =
+    Bytes.blit_string s 0 b.inner.buffer b.position len;
+  ) else
+    Bytes.unsafe_blit_string s 0 buffer position len;
   b.position <- new_position
```
(same here)

gasche commented on stdlib/buffer.ml:
```diff
@@ -334,33 +332,58 @@ external unsafe_set_int8 : bytes -> int -> int -> unit = "%bytes_unsafe_set"
+    set_int8 b.inner.buffer b.position x
+  ) else
+    unsafe_set_int8 buffer position x;
   b.position <- new_position
```
(same here)

gasche commented on stdlib/buffer.ml:
```diff
@@ -334,33 +332,58 @@ external unsafe_set_int8 : bytes -> int -> int -> unit = "%bytes_unsafe_set"
+    set_int16 b.inner.buffer b.position x
+  ) else
+    unsafe_set_int16 buffer position x;
   b.position <- new_position
```
(same here)

gasche commented on stdlib/buffer.ml:
```diff
@@ -334,33 +332,58 @@ external unsafe_set_int8 : bytes -> int -> int -> unit = "%bytes_unsafe_set"
+    set_int32 b.inner.buffer b.position x
+  ) else
+    unsafe_set_int32 buffer position x;
   b.position <- new_position
```
(same here)

gasche commented on stdlib/buffer.ml:
```diff
@@ -334,33 +332,58 @@ external unsafe_set_int8 : bytes -> int -> int -> unit = "%bytes_unsafe_set"
+    set_int64 b.inner.buffer b.position x
+  ) else
+    unsafe_set_int64 buffer position x;
   b.position <- new_position
```
(same here)

gasche commented on stdlib/buffer.ml:
```diff
@@ -87,102 +96,91 @@ let resize b more =
+  Bytes.blit b.inner.buffer 0 new_buffer 0 b.position;
+  b.inner <- { buffer = new_buffer; length = !new_len };
+  assert (b.position + more <= b.inner.length);
+  assert (old_pos + more <= b.inner.length);
```
I would get rid of the assert `b.position + more <= b.inner.length`. It was previously used to justify using unsafe writes in the `add_*` functions, but we cannot use it anymore as it may be now violated by races at any time. In fact I would just get rid of the two asserts.

Octachron replied:
Removing these two asserts that may fail randomly (due to a race) sounds like a good idea to me too (since `assert` are not good error messages).

gasche reviewed (commented):
This is a fine PR with good benchmarking. I suggest a slightly different approach to implement the `add_*` functions, with a recursive call instead of a checked write on resize. (This gives arguably-simpler code, and shorter.)

lthls commented on stdlib/buffer.ml:
```diff
@@ -87,102 +96,91 @@ let resize b more =
-  b.length <- !new_len;
-  assert (b.position + more <= b.length);
-  assert (old_pos + more <= b.length);
+  Bytes.blit b.inner.buffer 0 new_buffer 0 b.position;
```
I think that we could save the whole inner buffer at the start of this function, and use it here as the source instead of reloading. The length (`old_len` here) would also be used instead of `b.position` to specify the number of bytes to copy. This would avoid potential exceptions from `blit` when concurrent modification occurs, at the cost of being more likely to drop concurrently added data. We could even switch back to `unsafe_blit` in this case, although I would argue that it's better to be safe in this non-critical part of the code.

Octachron replied:
This sounds like a good idea in term of code regularity (and in term of regular behavior in presence of a data race).

Do you think it is really worth it to use `old_len` rather than `old_pos` with the aim of minimizing dropping unsynchronized data?

Any opinion on this "specification" change @gasche?

gasche replied:
I was also tempted to suggest to reuse `old_pos` instead of `b.position`, but I decided to not make the point explicitly -- you only get a *fraction* of my nitpick comments!

For me the problem with the assert is not that they fail (I think it's actually better to fail if we detect this kind of races), but that the code reads as if we believed that they will never fail. (You removed a lengthy comment that is mostly useless now, but it contained this useful information about the assertion that we know they may fail.)

It's actually fun to wonder whether we could detect all resize/resize races, fail with an error in this case, and behave somewhat-nicely -- at no performance cost. I suppose that suggestions similar to the "generation count" from the dynarray discussion could do it. But this could be left to a follow-up PR.

lthls replied:
If you stick to `b.inner.buffer` in the call to `blit` then you're already in a case where you don't know the length of the buffer, so re-fetching `b.position` is the solution least likely to drop concurrently added data (and most likely to be in sync with the actual buffer contents in general).
If you want to avoid out-of-bound accesses as much as possible, then blitting from the old buffer and `old_pos` is safer, even safer with `old_len` instead of `old_pos` (you guard against a reset occurring between fetching `b.position` and `b.inner`), but you're guaranteed to drop any concurrent updates occurring while you're computing the size and allocating the new array.

Octachron replied:
Then let's stick to just removing the assertions for this PR.

gasche reviewed (approved):
I reviewed the change and believe it is correct. There are some bits I would have written slightly differently (same for @lthls with different differences), but this is basically fine. I'm now eager to get this merged and move to some other exciting technical discussion.

Octachron commented:
I also agree that discussing further optimization could be done after 5.0 .

Octachron committed:
stdlib buffer review: remove misleading assertions

Those assertions in Buffer.resize were there to check the correctness of
invariants that relied on the global OCaml runtime lock. With OCaml 5,
those invariants cannot be meaningful enforced any longer. It seems
clearer to avoid giving the impression that those assertions cannot
fail.

Octachron committed:
fix: always reload position in the cold branch

Octachron committed:
fix: remove all asserts related to old invariants

Octachron commented:
I have pushed two minor "fixes":
- the first one ensures that we always reload the position in the cold path (that was the case everywhere except `add_char`).
- the second one remove one more assertion related to the old invariants.

gasche commented:
One code invariant that I would find reassuring is that we always access `b.inner.length` and `b.inner.buffer` together, instead of doing two separate reads. This is necessary for safety when performing unsafe writes, but it would be a good practice to follow in general. Currently this invariant is broken in the `add_utf*_uchar` functions and in `unsafe_add_channel_up_to `. Would you consider fixing this? It would also make the code more consistent.

Octachron commented:
I have the impression that the invariant is respected in those functions: I read
```ocaml
  let pos = b.position in
  if pos >= b.inner.length then resize b uchar_utf_8_byte_length_max;
  let n = Bytes.set_utf_8_uchar b.inner.buffer pos u in
```
as two reads of the inner buffer:
```ocaml
  let pos = b.position in
  let { length; _ } = b.inner in
  if pos >= length then resize b uchar_utf_8_byte_length_max;
  (* reload the possibly extended inner buffer *)
  let { buffer; _ } = b.inner in 
  let n = Bytes.set_utf_8_uchar buffer pos u in
```
Would you prefer to use that second form that makes the reads on the inner buffer more explicit?

gasche commented:
My understanding of your kind reply is that I misread those examples and that, indeed, the variant your propose (inspired by my suggestion) is slightly better, but this doesn't matter in the context of merging for 5.0.
